// @cyclonedx/cyclonedx-library/dist.node/builders/fromNodePackageJson.node.js
"use strict";
/*!
This file is part of CycloneDX JavaScript Library.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
Copyright (c) OWASP Foundation. All Rights Reserved.
*/
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _ToolBuilder_extRefFactory, _ComponentBuilder_extRefFactory, _ComponentBuilder_licenseFactory;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComponentBuilder = exports.ToolBuilder = void 0;
const packageJson_1 = require("../_helpers/packageJson");
const Enums = require("../enums");
const Models = require("../models");
class ToolBuilder {
    constructor(extRefFactory) {
        _ToolBuilder_extRefFactory.set(this, void 0);
        __classPrivateFieldSet(this, _ToolBuilder_extRefFactory, extRefFactory, "f");
    }
    get extRefFactory() {
        return __classPrivateFieldGet(this, _ToolBuilder_extRefFactory, "f");
    }
    makeTool(data) {
        const [name, vendor] = typeof data.name === 'string'
            ? (0, packageJson_1.splitNameGroup)(data.name)
            : [];
        return new Models.Tool({
            vendor,
            name,
            version: (typeof data.version === 'string')
                ? data.version
                : undefined,
            externalReferences: new Models.ExternalReferenceRepository(__classPrivateFieldGet(this, _ToolBuilder_extRefFactory, "f").makeExternalReferences(data))
        });
    }
}
exports.ToolBuilder = ToolBuilder;
_ToolBuilder_extRefFactory = new WeakMap();
class ComponentBuilder {
    constructor(extRefFactory, licenseFactory) {
        _ComponentBuilder_extRefFactory.set(this, void 0);
        _ComponentBuilder_licenseFactory.set(this, void 0);
        __classPrivateFieldSet(this, _ComponentBuilder_extRefFactory, extRefFactory, "f");
        __classPrivateFieldSet(this, _ComponentBuilder_licenseFactory, licenseFactory, "f");
    }
    get extRefFactory() {
        return __classPrivateFieldGet(this, _ComponentBuilder_extRefFactory, "f");
    }
    get licenseFactory() {
        return __classPrivateFieldGet(this, _ComponentBuilder_licenseFactory, "f");
    }
    makeComponent(data, type = Enums.ComponentType.Library) {
        if (typeof data.name !== 'string') {
            return undefined;
        }
        const [name, group] = (0, packageJson_1.splitNameGroup)(data.name);
        if (name.length <= 0) {
            return undefined;
        }
        const author = typeof data.author === 'string'
            ? data.author
            : (typeof data.author?.name === 'string'
                ? data.author.name
                : undefined);
        const description = typeof data.description === 'string'
            ? data.description
            : undefined;
        const version = typeof data.version === 'string'
            ? data.version
            : undefined;
        const externalReferences = __classPrivateFieldGet(this, _ComponentBuilder_extRefFactory, "f").makeExternalReferences(data);
        const licenses = new Models.LicenseRepository();
        if (typeof data.license === 'string') {
            licenses.add(__classPrivateFieldGet(this, _ComponentBuilder_licenseFactory, "f").makeFromString(data.license));
        }
        if (Array.isArray(data.licenses)) {
            for (const licenseData of data.licenses) {
                if (typeof licenseData?.type === 'string') {
                    const license = __classPrivateFieldGet(this, _ComponentBuilder_licenseFactory, "f").makeDisjunctive(licenseData.type);
                    license.url = typeof licenseData.url === 'string'
                        ? licenseData.url
                        : undefined;
                    licenses.add(license);
                }
            }
        }
        return new Models.Component(type, name, {
            author,
            description,
            externalReferences: new Models.ExternalReferenceRepository(externalReferences),
            group,
            licenses,
            version
        });
    }
}
exports.ComponentBuilder = ComponentBuilder;
_ComponentBuilder_extRefFactory = new WeakMap(), _ComponentBuilder_licenseFactory = new WeakMap();
