// @cyclonedx/cyclonedx-library/dist.node/models/bom.js
"use strict";
/*!
This file is part of CycloneDX JavaScript Library.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
Copyright (c) OWASP Foundation. All Rights Reserved.
*/
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _Bom_version, _Bom_serialNumber;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bom = void 0;
const types_1 = require("../types");
const component_1 = require("./component");
const metadata_1 = require("./metadata");
const vulnerability_1 = require("./vulnerability");
class Bom {
    constructor(op = {}) {
        _Bom_version.set(this, 1);
        _Bom_serialNumber.set(this, void 0);
        this.metadata = op.metadata ?? new metadata_1.Metadata();
        this.components = op.components ?? new component_1.ComponentRepository();
        this.version = op.version ?? this.version;
        this.vulnerabilities = op.vulnerabilities ?? new vulnerability_1.VulnerabilityRepository();
        this.serialNumber = op.serialNumber;
    }
    get version() {
        return __classPrivateFieldGet(this, _Bom_version, "f");
    }
    set version(value) {
        if (!(0, types_1.isPositiveInteger)(value)) {
            throw new TypeError('Not PositiveInteger');
        }
        __classPrivateFieldSet(this, _Bom_version, value, "f");
    }
    get serialNumber() {
        return __classPrivateFieldGet(this, _Bom_serialNumber, "f");
    }
    set serialNumber(value) {
        __classPrivateFieldSet(this, _Bom_serialNumber, value === ''
            ? undefined
            : value, "f");
    }
}
exports.Bom = Bom;
_Bom_version = new WeakMap(), _Bom_serialNumber = new WeakMap();
