// @cyclonedx/cyclonedx-library/dist.node/models/bomLink.js
"use strict";
/*!
This file is part of CycloneDX JavaScript Library.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
Copyright (c) OWASP Foundation. All Rights Reserved.
*/
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _BomLinkBase_value, _a, _BomLinkDocument_pattern, _b, _BomLinkElement_pattern;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BomLinkElement = exports.BomLinkDocument = void 0;
class BomLinkBase {
    constructor(value) {
        _BomLinkBase_value.set(this, void 0);
        this.value = value;
    }
    set value(value) {
        if (!this._isValid(value)) {
            throw new RangeError('invalid value');
        }
        __classPrivateFieldSet(this, _BomLinkBase_value, value, "f");
    }
    get value() {
        return __classPrivateFieldGet(this, _BomLinkBase_value, "f");
    }
    compare(other) {
        return this.toString().localeCompare(other.toString());
    }
    toString() {
        return this.value;
    }
}
_BomLinkBase_value = new WeakMap();
class BomLinkDocument extends BomLinkBase {
    static isValid(value) {
        return typeof value === 'string' &&
            __classPrivateFieldGet(this, _a, "f", _BomLinkDocument_pattern).test(value);
    }
    _isValid(value) {
        return _a.isValid(value);
    }
}
exports.BomLinkDocument = BomLinkDocument;
_a = BomLinkDocument;
_BomLinkDocument_pattern = { value: /^urn:cdx:[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\/[1-9][0-9]*$/ };
class BomLinkElement extends BomLinkBase {
    static isValid(value) {
        return typeof value === 'string' &&
            __classPrivateFieldGet(this, _b, "f", _BomLinkElement_pattern).test(value);
    }
    _isValid(value) {
        return _b.isValid(value);
    }
}
exports.BomLinkElement = BomLinkElement;
_b = BomLinkElement;
_BomLinkElement_pattern = { value: /^urn:cdx:[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\/[1-9][0-9]*#.+$/ };
