// @cyclonedx/cyclonedx-library/dist.node/models/component.js
"use strict";
/*!
This file is part of CycloneDX JavaScript Library.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
Copyright (c) OWASP Foundation. All Rights Reserved.
*/
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Component_bomRef, _Component_cpe;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComponentEvidence = exports.ComponentRepository = exports.Component = void 0;
const sortable_1 = require("../_helpers/sortable");
const tree_1 = require("../_helpers/tree");
const types_1 = require("../types");
const bomRef_1 = require("./bomRef");
const externalReference_1 = require("./externalReference");
const hash_1 = require("./hash");
const license_1 = require("./license");
const property_1 = require("./property");
class Component {
    constructor(type, name, op = {}) {
        _Component_bomRef.set(this, void 0);
        _Component_cpe.set(this, void 0);
        __classPrivateFieldSet(this, _Component_bomRef, new bomRef_1.BomRef(op.bomRef), "f");
        this.type = type;
        this.name = name;
        this.supplier = op.supplier;
        this.author = op.author;
        this.copyright = op.copyright;
        this.externalReferences = op.externalReferences ?? new externalReference_1.ExternalReferenceRepository();
        this.group = op.group;
        this.hashes = op.hashes ?? new hash_1.HashDictionary();
        this.licenses = op.licenses ?? new license_1.LicenseRepository();
        this.publisher = op.publisher;
        this.purl = op.purl;
        this.scope = op.scope;
        this.swid = op.swid;
        this.version = op.version;
        this.description = op.description;
        this.dependencies = op.dependencies ?? new bomRef_1.BomRefRepository();
        this.components = op.components ?? new ComponentRepository();
        this.cpe = op.cpe;
        this.properties = op.properties ?? new property_1.PropertyRepository();
        this.evidence = op.evidence;
    }
    get bomRef() {
        return __classPrivateFieldGet(this, _Component_bomRef, "f");
    }
    get cpe() {
        return __classPrivateFieldGet(this, _Component_cpe, "f");
    }
    set cpe(value) {
        if (value !== undefined && !(0, types_1.isCPE)(value)) {
            throw new TypeError('Not CPE nor undefined');
        }
        __classPrivateFieldSet(this, _Component_cpe, value, "f");
    }
    compare(other) {
        const bomRefCompare = this.bomRef.compare(other.bomRef);
        if (bomRefCompare !== 0) {
            return bomRefCompare;
        }
        if (this.purl !== undefined && other.purl !== undefined) {
            return this.purl.toString().localeCompare(other.purl.toString());
        }
        if (__classPrivateFieldGet(this, _Component_cpe, "f") !== undefined && __classPrivateFieldGet(other, _Component_cpe, "f") !== undefined) {
            return __classPrivateFieldGet(this, _Component_cpe, "f").localeCompare(__classPrivateFieldGet(other, _Component_cpe, "f"));
        }
        return (this.group ?? '').localeCompare(other.group ?? '') ||
            this.name.localeCompare(other.name) ||
            (this.version ?? '').localeCompare(other.version ?? '');
    }
}
exports.Component = Component;
_Component_bomRef = new WeakMap(), _Component_cpe = new WeakMap();
class ComponentRepository extends sortable_1.SortableComparables {
    *[tree_1.treeIteratorSymbol]() {
        for (const component of this) {
            yield component;
            yield* component.components[tree_1.treeIteratorSymbol]();
        }
    }
}
exports.ComponentRepository = ComponentRepository;
class ComponentEvidence {
    constructor(op = {}) {
        this.licenses = op.licenses ?? new license_1.LicenseRepository();
        this.copyright = op.copyright ?? new sortable_1.SortableStringables();
    }
}
exports.ComponentEvidence = ComponentEvidence;
