// @cyclonedx/cyclonedx-library/dist.node/models/license.js
"use strict";
/*!
This file is part of CycloneDX JavaScript Library.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
Copyright (c) OWASP Foundation. All Rights Reserved.
*/
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _LicenseExpression_expression, _DisjunctiveLicenseBase_url, _SpdxLicense_id, _a, _LicenseRepository_compareItems;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LicenseRepository = exports.SpdxLicense = exports.NamedLicense = exports.LicenseExpression = void 0;
class LicenseExpression {
    constructor(expression) {
        _LicenseExpression_expression.set(this, void 0);
        this.expression = expression;
    }
    get expression() {
        return __classPrivateFieldGet(this, _LicenseExpression_expression, "f");
    }
    set expression(value) {
        if (value === '') {
            throw new RangeError('value is empty string');
        }
        __classPrivateFieldSet(this, _LicenseExpression_expression, value, "f");
    }
    compare(other) {
        return __classPrivateFieldGet(this, _LicenseExpression_expression, "f").localeCompare(__classPrivateFieldGet(other, _LicenseExpression_expression, "f"));
    }
}
exports.LicenseExpression = LicenseExpression;
_LicenseExpression_expression = new WeakMap();
class DisjunctiveLicenseBase {
    constructor(op = {}) {
        _DisjunctiveLicenseBase_url.set(this, void 0);
        this.text = op.text;
        this.url = op.url;
    }
    get url() {
        return __classPrivateFieldGet(this, _DisjunctiveLicenseBase_url, "f");
    }
    set url(value) {
        __classPrivateFieldSet(this, _DisjunctiveLicenseBase_url, value === ''
            ? undefined
            : value, "f");
    }
}
_DisjunctiveLicenseBase_url = new WeakMap();
class NamedLicense extends DisjunctiveLicenseBase {
    constructor(name, op = {}) {
        super(op);
        this.name = name;
    }
    compare(other) {
        return this.name.localeCompare(other.name);
    }
}
exports.NamedLicense = NamedLicense;
class SpdxLicense extends DisjunctiveLicenseBase {
    constructor(id, op = {}) {
        super(op);
        _SpdxLicense_id.set(this, void 0);
        this.id = id;
    }
    get id() {
        return __classPrivateFieldGet(this, _SpdxLicense_id, "f");
    }
    set id(value) {
        if (value === '') {
            throw new RangeError('value is empty string');
        }
        __classPrivateFieldSet(this, _SpdxLicense_id, value, "f");
    }
    compare(other) {
        return __classPrivateFieldGet(this, _SpdxLicense_id, "f").localeCompare(__classPrivateFieldGet(other, _SpdxLicense_id, "f"));
    }
}
exports.SpdxLicense = SpdxLicense;
_SpdxLicense_id = new WeakMap();
class LicenseRepository extends Set {
    sorted() {
        return Array.from(this).sort(__classPrivateFieldGet(_a, _a, "m", _LicenseRepository_compareItems));
    }
}
exports.LicenseRepository = LicenseRepository;
_a = LicenseRepository, _LicenseRepository_compareItems = function _LicenseRepository_compareItems(a, b) {
    if (a.constructor === b.constructor) {
        return a.compare(b);
    }
    return a.constructor.name.localeCompare(b.constructor.name);
};
