// @cyclonedx/cyclonedx-library/dist.node/models/lifecycle.js
"use strict";
/*!
This file is part of CycloneDX JavaScript Library.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
Copyright (c) OWASP Foundation. All Rights Reserved.
*/
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _a, _LifecycleRepository_compareItems;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LifecycleRepository = exports.NamedLifecycle = void 0;
class NamedLifecycle {
    constructor(name, op = {}) {
        this.name = name;
        this.description = op.description;
    }
    compare(other) {
        return this.name.localeCompare(other.name);
    }
}
exports.NamedLifecycle = NamedLifecycle;
class LifecycleRepository extends Set {
    sorted() {
        return Array.from(this).sort(__classPrivateFieldGet(_a, _a, "m", _LifecycleRepository_compareItems));
    }
}
exports.LifecycleRepository = LifecycleRepository;
_a = LifecycleRepository, _LifecycleRepository_compareItems = function _LifecycleRepository_compareItems(a, b) {
    if (a.constructor === b.constructor) {
        return a instanceof NamedLifecycle
            ? a.compare(b)
            : a.localeCompare(b);
    }
    return a.constructor.name.localeCompare(b.constructor.name);
};
