// @cyclonedx/cyclonedx-library/dist.node/models/metadata.js
"use strict";
/*!
This file is part of CycloneDX JavaScript Library.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
Copyright (c) OWASP Foundation. All Rights Reserved.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.Metadata = void 0;
const license_1 = require("./license");
const lifecycle_1 = require("./lifecycle");
const organizationalContact_1 = require("./organizationalContact");
const property_1 = require("./property");
const tool_1 = require("./tool");
class Metadata {
    constructor(op = {}) {
        this.timestamp = op.timestamp;
        this.lifecycles = op.lifecycles ?? new lifecycle_1.LifecycleRepository();
        this.tools = op.tools ?? new tool_1.ToolRepository();
        this.authors = op.authors ?? new organizationalContact_1.OrganizationalContactRepository();
        this.component = op.component;
        this.manufacture = op.manufacture;
        this.supplier = op.supplier;
        this.licenses = op.licenses ?? new license_1.LicenseRepository();
        this.properties = op.properties ?? new property_1.PropertyRepository();
    }
}
exports.Metadata = Metadata;
