// @cyclonedx/cyclonedx-library/dist.node/models/swid.js
"use strict";
/*!
This file is part of CycloneDX JavaScript Library.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
Copyright (c) OWASP Foundation. All Rights Reserved.
*/
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _SWID_tagVersion;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SWID = void 0;
const types_1 = require("../types");
class SWID {
    constructor(tagId, name, op = {}) {
        _SWID_tagVersion.set(this, void 0);
        this.tagId = tagId;
        this.name = name;
        this.version = op.version;
        this.patch = op.patch;
        this.text = op.text;
        this.url = op.url;
        this.tagVersion = op.tagVersion;
    }
    get tagVersion() {
        return __classPrivateFieldGet(this, _SWID_tagVersion, "f");
    }
    set tagVersion(value) {
        if (value !== undefined && !(0, types_1.isNonNegativeInteger)(value)) {
            throw new TypeError('Not NonNegativeInteger nor undefined');
        }
        __classPrivateFieldSet(this, _SWID_tagVersion, value, "f");
    }
}
exports.SWID = SWID;
_SWID_tagVersion = new WeakMap();
