// @cyclonedx/cyclonedx-library/dist.node/validation/xmlValidator.node.js
"use strict";
/*!
This file is part of CycloneDX JavaScript Library.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
Copyright (c) OWASP Foundation. All Rights Reserved.
*/
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _XmlValidator_instances, _XmlValidator_schema, _XmlValidator_getSchemaFile, _XmlValidator_getSchema;
Object.defineProperty(exports, "__esModule", { value: true });
exports.XmlValidator = void 0;
const promises_1 = require("fs/promises");
const url_1 = require("url");
const resources_node_1 = require("../resources.node");
const baseValidator_1 = require("./baseValidator");
const errors_1 = require("./errors");
let _parser;
async function getParser() {
    if (_parser === undefined) {
        let libxml;
        try {
            libxml = await Promise.resolve().then(() => require('libxmljs2'));
        }
        catch (err) {
            throw new errors_1.MissingOptionalDependencyError('No XML validator available.' +
                ' Please install the optional dependency "libxmljs2".' +
                ' Please make sure the system meets the requirements for "node-gyp", see https://github.com/TooTallNate/node-gyp#installation', err);
        }
        _parser = libxml.parseXml;
    }
    return _parser;
}
const xmlParseOptions = Object.freeze({
    nonet: true,
    compact: true
});
class XmlValidator extends baseValidator_1.BaseValidator {
    constructor() {
        super(...arguments);
        _XmlValidator_instances.add(this);
        _XmlValidator_schema.set(this, void 0);
    }
    async validate(data) {
        const [parse, schema] = await Promise.all([
            getParser(),
            __classPrivateFieldGet(this, _XmlValidator_instances, "m", _XmlValidator_getSchema).call(this)
        ]);
        const doc = parse(data, xmlParseOptions);
        return doc.validate(schema)
            ? null
            : doc.validationErrors;
    }
}
exports.XmlValidator = XmlValidator;
_XmlValidator_schema = new WeakMap(), _XmlValidator_instances = new WeakSet(), _XmlValidator_getSchemaFile = function _XmlValidator_getSchemaFile() {
    return resources_node_1.FILES.CDX.XML_SCHEMA[this.version];
}, _XmlValidator_getSchema = async function _XmlValidator_getSchema() {
    if (undefined === __classPrivateFieldGet(this, _XmlValidator_schema, "f")) {
        const file = __classPrivateFieldGet(this, _XmlValidator_instances, "m", _XmlValidator_getSchemaFile).call(this);
        if (file === undefined) {
            throw new errors_1.NotImplementedError(this.version);
        }
        const [parse, schema] = await Promise.all([
            getParser(),
            (0, promises_1.readFile)(file, 'utf-8')
        ]);
        __classPrivateFieldSet(this, _XmlValidator_schema, parse(schema, { ...xmlParseOptions, baseUrl: (0, url_1.pathToFileURL)(file).toString() }), "f");
    }
    return __classPrivateFieldGet(this, _XmlValidator_schema, "f");
};
